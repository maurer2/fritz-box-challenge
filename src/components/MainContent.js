import React from 'react';
import styled from 'styled-components/macro';
import { CSSTransitionGroup } from 'react-transition-group';
import PropTypes from 'prop-types';

const MainWrapper = styled.main`
  position: relative;
  display: block;
  width: 100%;

  .slide-leave {
    position: relative;
    transform: translateX(0);
    transition: transform 500ms;

    &.slide-leave-active {
      position: absolute;
      top: 0;
      left: 0;
      right: 0;
      transform: translateX(-100%);
    }
  }

  .slide-enter {
    position: absolute;
    top: 0;
    left: 0;
    right: 0;
    transform: translateX(100%);
    transition: transform 500ms;

    &.slide-enter-active {
      position: relative;
      transform: translateX(0);
    }
  }
`;

const MainContent = (props) => {
  const { isUpdating } = props;

  return (
    <MainWrapper>
      <CSSTransitionGroup
        component={ React.Fragment }
        transitionName="slide"
        transitionEnterTimeout={ 500 }
        transitionLeaveTimeout={ 500 }
        transitionLeave={ true }
      >
        {!isUpdating && props.children}
      </CSSTransitionGroup>
    </MainWrapper>
  );
};

MainContent.propTypes = {
  isUpdating: PropTypes.bool,
  children: PropTypes.array,
};

export default MainContent;
