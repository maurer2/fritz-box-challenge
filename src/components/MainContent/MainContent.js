import React from 'react';
import styled from 'styled-components/macro';
import { CSSTransitionGroup } from 'react-transition-group';
import PropTypes from 'prop-types';

const MainWrapper = styled.main`
  width: 100%;

  > div {
    position: absolute;
    top: 50%;
    left: 0;
    right: 0;
    transform: translateX(0) translateY(-50%);
  }

  // slide in right
  > .slide-in-right-leave {
    opacity: 1;
    transform: translateX(0) translateY(-50%);
    transition: transform 500ms, opacity 500ms;

    &.slide-in-right-leave-active {
      opacity: 0.01;
      transform: translateX(-100%) translateY(-50%);
    }
  }

  > .slide-in-right-enter {
    opacity: 0.01;
    transform: translateX(100%) translateY(-50%);
    transition: transform 500ms, opacity 500ms;

    &.slide-in-right-enter-active {
      opacity: 1;
      transform: translateX(0) translateY(-50%);
    }
  }

  > .slide-in-right-appear {
    opacity: 0.01;
    transition: opacity 1000ms;

    &.slide-in-right-appear-active {
      opacity: 1;
    }
  }

  // slide in left
  > .slide-in-left-leave {
    opacity: 1;
    transform: translateX(0) translateY(-50%);
    transition: transform 500ms, opacity 500ms;

    &.slide-in-left-leave-active {
      opacity: 0.01;
      transform: translateX(100%) translateY(-50%);
    }
  }

  > .slide-in-left-enter {
    opacity: 0.01;
    transform: translateX(-100%) translateY(-50%);
    transition: transform 500ms, opacity 500ms;

    &.slide-in-left-enter-active {
      opacity: 1;
      transform: translateX(0) translateY(-50%);
    }
  }

  > .slide-in-left-appear {
    opacity: 0.01;
    transition: opacity 1000ms;

    &.slide-in-left-appear-active {
      opacity: 1;
    }
  }
`;

const MainContent = ({ handleClick, currentIndex, oldIndex, children }) => {
  const slideInFromRight = currentIndex > oldIndex;

  return (
    <MainWrapper onClick={ handleClick }>
      <CSSTransitionGroup
        component={ React.Fragment }
        transitionName={ slideInFromRight ? 'slide-in-right' : 'slide-in-left' }
        transitionEnterTimeout={ 500 }
        transitionLeaveTimeout={ 500 }
        transitionAppearTimeout={ 100 }
        transitionLeave={ true }
        transitionAppear={ true }
      >
        { children }
      </CSSTransitionGroup>
    </MainWrapper>
  );
};

MainContent.propTypes = {
  currentIndex: PropTypes.number.isRequired,
  oldIndex: PropTypes.number.isRequired,
  children: PropTypes.object.isRequired,
  handleClick: PropTypes.func.isRequired,
};

export { MainContent };
