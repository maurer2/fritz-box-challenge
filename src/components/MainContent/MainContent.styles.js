import styled from 'styled-components/macro';

export const MainWrapper = styled.main`
  width: 100%;

  > div {
    position: absolute;
    top: 50%;
    left: 0;
    right: 0;
    transform: translateX(0) translateY(-50%);
  }

  // slide in right
  > .slide-in-right-leave {
    opacity: 1;
    transform: translateX(0) translateY(-50%);
    transition: transform 500ms, opacity 500ms;

    &.slide-in-right-leave-active {
      opacity: 0.01;
      transform: translateX(-100%) translateY(-50%);
    }
  }

  > .slide-in-right-enter {
    opacity: 0.01;
    transform: translateX(100%) translateY(-50%);
    transition: transform 500ms, opacity 500ms;

    &.slide-in-right-enter-active {
      opacity: 1;
      transform: translateX(0) translateY(-50%);
    }
  }

  > .slide-in-right-appear {
    opacity: 0.01;
    transition: opacity 1000ms;

    &.slide-in-right-appear-active {
      opacity: 1;
    }
  }

  // slide in left
  > .slide-in-left-leave {
    opacity: 1;
    transform: translateX(0) translateY(-50%);
    transition: transform 500ms, opacity 500ms;

    &.slide-in-left-leave-active {
      opacity: 0.01;
      transform: translateX(100%) translateY(-50%);
    }
  }

  > .slide-in-left-enter {
    opacity: 0.01;
    transform: translateX(-100%) translateY(-50%);
    transition: transform 500ms, opacity 500ms;

    &.slide-in-left-enter-active {
      opacity: 1;
      transform: translateX(0) translateY(-50%);
    }
  }

  > .slide-in-left-appear {
    opacity: 0.01;
    transition: opacity 1000ms;

    &.slide-in-left-appear-active {
      opacity: 1;
    }
  }
`;
