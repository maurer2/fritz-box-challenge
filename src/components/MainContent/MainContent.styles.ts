import styled from 'styled-components';
import { type TransitionStatus } from 'react-transition-state';

export interface MainWrapperStyleProps {
  $status: TransitionStatus;
}

export const MainWrapperSlideWrapper = styled.main<MainWrapperStyleProps>`
  position: absolute;
  top: 50%;
  right: 0;
  left: 0;
  transform: translateX(0) translateY(-50%);
  background: red;

  ${({ $status }) => ($status === 'entering') && `
      background: green;
    `}

    ${({ $status }) => ($status === 'entered') && `
      background: blue;
    `}

    ${({ $status }) => ($status === 'exiting') && `
      background: yellow;
    `}

    ${({ $status }) => ($status === 'exited') && `
      background: purple;
    `}
`;

export const MainWrapper = styled.main`
  position: relative;
  grid-area: main;
  width: 100%;

  > div {
    position: absolute;
    top: 50%;
    right: 0;
    left: 0;
    transform: translateX(0) translateY(-50%);
  }

  // slide in right
  > .slide-in-right-leave {
    transform: translateX(0) translateY(-50%);
    opacity: 1;
    transition: transform 500ms, opacity 500ms;

    &.slide-in-right-leave-active {
      transform: translateX(-100%) translateY(-50%);
      opacity: 0.01;
    }
  }

  > .slide-in-right-enter {
    transform: translateX(100%) translateY(-50%);
    opacity: 0.01;
    transition: transform 500ms, opacity 500ms;

    &.slide-in-right-enter-active {
      transform: translateX(0) translateY(-50%);
      opacity: 1;
    }
  }

  > .slide-in-right-appear {
    opacity: 0.01;
    transition: opacity 1000ms;

    &.slide-in-right-appear-active {
      opacity: 1;
    }
  }

  // slide in left
  > .slide-in-left-leave {
    transform: translateX(0) translateY(-50%);
    opacity: 1;
    transition: transform 500ms, opacity 500ms;

    &.slide-in-left-leave-active {
      transform: translateX(100%) translateY(-50%);
      opacity: 0.01;
    }
  }

  > .slide-in-left-enter {
    transform: translateX(-100%) translateY(-50%);
    opacity: 0.01;
    transition: transform 500ms, opacity 500ms;

    &.slide-in-left-enter-active {
      transform: translateX(0) translateY(-50%);
      opacity: 1;
    }
  }

  > .slide-in-left-appear {
    opacity: 0.01;
    transition: opacity 1000ms;

    &.slide-in-left-appear-active {
      opacity: 1;
    }
  }
`;

export const MainWrapperTrigger = styled.button`
  position: absolute;
  inset: 0;
  padding: 0;
  margin: 0;
  background: transparent;
  cursor: pointer;
  border: none;

  > .text {
    position: absolute;
    width: 1px;
    height: 1px;
    padding: 0;
    margin: -1px;
    overflow: hidden;
    clip: rect(0, 0, 0, 0);
    white-space: nowrap;
    border-width: 0;
  }
`;
