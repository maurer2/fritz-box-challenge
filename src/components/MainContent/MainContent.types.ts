export interface StringMap {
  [key: string]: string;
}

export interface MainContentProps {
  handleClick: React.EventHandler<any>;
  currentIndex: number;
  oldIndex: number;
  children: React.ReactNode;
}
