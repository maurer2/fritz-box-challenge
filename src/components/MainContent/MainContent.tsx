import React, { useContext } from 'react';
import { CSSTransitionGroup } from 'react-transition-group'; // todo replace with https://github.com/szhsin/react-transition-state

import { Slide } from '../Slide';
import { BoxDataContext } from '../DataProvider';

import * as Styles from './MainContent.styles';

const MainContent = () => {
  const state = useContext(BoxDataContext);

  const {
    visibleComponents,
    boxData,
    currentIndex,
    prevIndex,
    updateCurrentIndex,
  } = state;

  if (!boxData) {
    return null;
  }

  const title = visibleComponents[currentIndex];
  const text = boxData?.[title];
  const slideInFromRight = currentIndex > prevIndex;

  function handleClick(): void {
    const indexOfLastEntry = visibleComponents.length - 1;
    const newCurrentIndex = currentIndex < indexOfLastEntry ? currentIndex + 1 : 0;

    updateCurrentIndex(newCurrentIndex);
  }

  return (
    <Styles.MainWrapper onClick={(): void => handleClick()}>
      <CSSTransitionGroup
        component={React.Fragment}
        transitionName={slideInFromRight ? 'slide-in-right' : 'slide-in-left'}
        transitionEnterTimeout={500}
        transitionLeaveTimeout={500}
        transitionAppearTimeout={100}
        transitionLeave
        transitionAppear
      >
        <Slide title={title} text={text} key={title} />
      </CSSTransitionGroup>
    </Styles.MainWrapper>
  );
};

export { MainContent };
