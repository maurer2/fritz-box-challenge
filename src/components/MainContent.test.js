import React from 'react';
import Enzyme, { shallow } from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';

import MainContent from './MainContent';

// React 16 Enzyme adapter
Enzyme.configure({ adapter: new Adapter() });

describe('MainContent', () => {
  const wrapper = shallow(<MainContent isUpdating={ false } />);

  it('should match snapshot', () => {
    expect(wrapper).toMatchSnapshot();
  });
});
