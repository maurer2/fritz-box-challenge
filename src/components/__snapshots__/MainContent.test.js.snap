// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MainContent should match snapshot 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <MainContent
    isUpdating={false}
  />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "checkPropTypes": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateError": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "function",
    "props": Object {
      "children": <CSSTransitionGroup
        component={Symbol(react.fragment)}
        transitionAppear={false}
        transitionEnter={true}
        transitionEnterTimeout={500}
        transitionLeave={true}
        transitionLeaveTimeout={500}
        transitionName="slide"
      />,
    },
    "ref": null,
    "rendered": Object {
      "instance": null,
      "key": undefined,
      "nodeType": "class",
      "props": Object {
        "children": undefined,
        "component": Symbol(react.fragment),
        "transitionAppear": false,
        "transitionEnter": true,
        "transitionEnterTimeout": 500,
        "transitionLeave": true,
        "transitionLeaveTimeout": 500,
        "transitionName": "slide",
      },
      "ref": null,
      "rendered": null,
      "type": [Function],
    },
    "type": Object {
      "$$typeof": Symbol(react.forward_ref),
      "attrs": Array [],
      "componentStyle": ComponentStyle {
        "componentId": "MainWrapper-sc-1d2488w-0",
        "isStatic": true,
        "rules": Array [
          "position:relative;display:block;width:100%;.slide-leave{position:relative;transform:translateX(0);transition:transform 500ms;&.slide-leave-active{position:absolute;top:0;left:0;right:0;transform:translateX(-100%);}}.slide-enter{position:absolute;top:0;left:0;right:0;transform:translateX(100%);transition:transform 500ms;&.slide-enter-active{position:relative;transform:translateX(0);}}",
        ],
      },
      "displayName": "MainWrapper",
      "foldedComponentIds": Array [],
      "render": [Function],
      "styledComponentId": "MainWrapper-sc-1d2488w-0",
      "target": "main",
      "toString": [Function],
      "warnTooManyClasses": [Function],
      "withComponent": [Function],
    },
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "function",
      "props": Object {
        "children": <CSSTransitionGroup
          component={Symbol(react.fragment)}
          transitionAppear={false}
          transitionEnter={true}
          transitionEnterTimeout={500}
          transitionLeave={true}
          transitionLeaveTimeout={500}
          transitionName="slide"
        />,
      },
      "ref": null,
      "rendered": Object {
        "instance": null,
        "key": undefined,
        "nodeType": "class",
        "props": Object {
          "children": undefined,
          "component": Symbol(react.fragment),
          "transitionAppear": false,
          "transitionEnter": true,
          "transitionEnterTimeout": 500,
          "transitionLeave": true,
          "transitionLeaveTimeout": 500,
          "transitionName": "slide",
        },
        "ref": null,
        "rendered": null,
        "type": [Function],
      },
      "type": Object {
        "$$typeof": Symbol(react.forward_ref),
        "attrs": Array [],
        "componentStyle": ComponentStyle {
          "componentId": "MainWrapper-sc-1d2488w-0",
          "isStatic": true,
          "rules": Array [
            "position:relative;display:block;width:100%;.slide-leave{position:relative;transform:translateX(0);transition:transform 500ms;&.slide-leave-active{position:absolute;top:0;left:0;right:0;transform:translateX(-100%);}}.slide-enter{position:absolute;top:0;left:0;right:0;transform:translateX(100%);transition:transform 500ms;&.slide-enter-active{position:relative;transform:translateX(0);}}",
          ],
        },
        "displayName": "MainWrapper",
        "foldedComponentIds": Array [],
        "render": [Function],
        "styledComponentId": "MainWrapper-sc-1d2488w-0",
        "target": "main",
        "toString": [Function],
        "warnTooManyClasses": [Function],
        "withComponent": [Function],
      },
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "legacyContextMode": "parent",
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getChildContext": Object {
            "calledByRenderer": false,
          },
          "getDerivedStateFromProps": true,
          "getSnapshotBeforeUpdate": true,
          "setState": Object {
            "skipsComponentDidUpdateOnNullish": true,
          },
        },
      },
    },
  },
}
`;
